app/components/app_activity_log_component.rb

```diff
@@ -302,60 +302,105 @@ def programmes_by_id
-          .flat_map { programmes_for(it) }
-          .reject { vaccinated_programmes.include?(it) }
-
-      expired_items = []
-      if programmes_with_expired_consents.any?
-        expired_items += ["consent", "health information"]
-      end
-      expired_items << "triage outcome" if programmes_with_expired_triages.any?
-      expired_items << "PSD status" if programmes_with_expired_psds.any?
+      vaccinated_but_seasonal_programmes =
+        all_programmes.select do |programme|
+          patient.vaccination_status(programme:, academic_year:).vaccinated? &&
+            programme.seasonal?
+        end
 
-      next if expired_items.empty?
+      expired_items = {
+        vaccinated_but_seasonal:
+          get_expired_items(
+            academic_year:,
+            programmes: vaccinated_but_seasonal_programmes
+          ),
+        not_vaccinated:
+          get_expired_items(
+            academic_year:,
+            programmes: not_vaccinated_programmes
+          )
+      }
 
-      programmes_with_expired_items = [
-        programmes_with_expired_consents,
-        programmes_with_expired_triages,
-        programmes_with_expired_psds
-      ].flatten.uniq
+      %i[not_vaccinated vaccinated_but_seasonal].each do |category|
+        expired_item_names = []
+        if expired_items[category][:consents].any?
+          expired_item_names += ["consent", "health information"]
+        end
+        if expired_items[category][:triages].any?
+          expired_item_names << "triage outcome"
+        end
+        if expired_items[category][:psds].any?
+          expired_item_names << "PSD status"
+        end
 
-      expired_items_sentence =
-        expired_items.to_sentence(
-          words_connector: ", ",
-          last_word_connector: " and "
-        )
+        next if expired_item_names.empty?
+
+        title =
+          "#{
+            expired_item_names.to_sentence(
+              words_connector: ", ",
+              last_word_connector: " and "
+            ).upcase_first
+          } expired"
+
+        body =
+          case category
+          when :not_vaccinated
+            "#{@patient.full_name} was not vaccinated."
+          when :vaccinated_but_seasonal
+            "#{@patient.full_name} was vaccinated."
+          end
 
-      {
-        title: "#{expired_items_sentence.upcase_first} expired",
-        body: "#{@patient.full_name} was not vaccinated.",
-        at: academic_year.to_academic_year_date_range.end.end_of_day - 1.second,
-        programmes: programmes_with_expired_items
-      }
+        programmes = expired_items[category].values.flatten.uniq
+
+        events << {
+          title:,
+          body:,
+          at:
+            academic_year.to_academic_year_date_range.end.end_of_day - 1.second,
+          programmes:
+        }
+      end
     end
+
+    events
+  end
+
+  private
+
+  def get_expired_items(academic_year:, programmes:)
```

Comment:
This is a nice refactor, but I think we could go further given that each one has the same three lines. Something like:

```rb
def filter_expired(objects, academic_year:, programmes:)
  objects
    .select { it.academic_year == academic_year }
    .flat_map { programmes_for(it) }
    .select { programmes.include?(it) }
end

def expired_objects(academic_year:, programmes:)
  {
    consents: filter_expired(consents, academic_year:, programmes:),
    triages: filter_expired(triages, academic_year:, programmes:),
    # etc
  }
end
```